### DISCUSSION TITLE
Orval in the age of React SSR (specifically NextJS14)

### DISCUSSION THREAD

**Discussion by devraj** [orval-labs/orval#1057] [General]
**Background**

With the massive changes happening around the Frontend Cloud and frameworks like NextJS rapidly moving the entire frontend world to server side rendering, many of the assumptions that frameworks like `axios` are at odds with how the hybrid rendering model works. For example React contexts and ideas registering `interceptors` and `defaults` in frameworks like `axios` don't work (at least as smoothly) as they did in the client side world.

This is mostly due to lifecycle issues that server side rendering poses. While frameworks like `nextjs` (at least in `14.0.3-canary`) provide features like `instrumentationHook` where one can run bootstrap code on startup, it still does not provide the same seamless setup of the axios client as you would get in a pure client side experience.

NextJS 14 in particular is also building upon the native [fetch api](https://nextjs.org/docs/app/api-reference/functions/fetch) with extensions for caching, invalidation, etc.

**Point for discussion**

Orval supports writing [custom http client](https://orval.dev/guides/custom-client) which the generator then uses instead of `axios` or I assume it is possible to write a custom `builder` just like the ones that are supported out of the box.

While I am a fan of `axios` and similar frameworks, I personally find `axios` an unnecessary layer when `fetch` and other native frameworks are way more mature than they have been in the past. Also with frameworks like Next building right on top of the native web frameworks it would make sense to get close to the metal (so to speak) as possible.

At the moment I have been working around `axios` behaviour with setup and teardown helpers because `interceptors` and `defaults` don't work so well for server side functions.

My discussion here is to explore:
- What plans orval has for the brave new world of SSR
- Would the core team consider implementing a fetch (optionally swr support) based builder (reason I ask is when I implement a custom http client, I lose a few nice things like passing in optional headers)
- orval's own implementation of things like providing global defaults (not sure if this is possible)
- Provide documentation on how to write more featured custom http clients (in case all this is possible by writing a custom http client)

> Note that [as a sponsor](https://github.com/anomaly) of this project, we would be willing to have a commercial discussion around building some of these features for the project.

**Comment by melloware** [comment]
I know prefetch functions were just added in 6.20.0 to support Next SSR: https://github.com/anymaniax/orval/pull/956

**Comment by Gustutu** [comment]
how about sveltekit ? I am having some trouble using orval with sveltekit for now.

**Comment by razzeee** [comment]
Would you be able to share your custom client?

**Comment by harunzafer** [comment]
SPA is much better and simpler in my opinion. I am using FastApi backend and SvelteKit as SPA and using Orval in between. [Here ](https://turtledev.io/blog/fastapi-tutorial-3-sveltekit-spa-setup-orval) is my setup.
